### DISCUSSION TITLE
Unable to call other tasks from test command

### DISCUSSION THREAD

**Discussion by ex0ns** [com-lihaoyi/mill#2603] [Q&A] [answered]
In a pretty simple build definition, I was trying to override the `test` method of a module, and I don't understand why what I am trying to perform is not valid:

```scala

import mill._
import scalalib._


object core extends ScalaModule {
  def scalaVersion = "2.13.11"

  object test extends ScalaTests with TestModule.ScalaTest {
    def other = T {
      T.log.info("Other log")
    }

    def test(args: String*) = T.command {
      other()
      T.log.info("Running test")
      super.test()
    }
  }
}

```
```
[error] /Users/ex0ns/tmp/build.sc:22:12: Target#apply() can only be used with a T{...} block
[error]       other()
[error]            ^
[error] /Users/ex0ns/tmp/build.sc:23:9: Target.ctx() / T.ctx() / T.* APIs can only be used with a T{...} block
[error]       T.log.info("Running test")
[error]         ^
[error] two errors found
```

whereas if I defined another target, with the same signature, it works fine:

```scala
def mytest(args: String*) = T.command { 
  T.log.info("test")
}
```

**Comment by lefou** [comment, marked as answer]
The issue with your attempt is, that you don't depend on the result of the `super.test()` command, but instead return the command itself. Or in other words, you forgot the second `()`.

```diff
 def test(args: String*) = T.command {
       other()
       T.log.info("Running test")
-      super.test()
+      super.test()()
     }
```

As `T.command` accepts a `Task` and you already return a task, you essentially return that `super.test` task and also access the `T.log.info` outside of a task, which is not possible. Hence, the somewhat unhelpful error message.

To use a task's outcome, you always add one addition parenthesis pair more than on the definition side.
This is the short for `task.apply()`. For commands, that don't need any arguments, that means you have two parenthesis pairs. The first holding the (empty) command arguments, and the second actually refering the the result value of that command.

**Reply by lefou** [reply]
Please note, that the `T.log.info("Running test")` will always run after the `super.test`, due to the fact that the `test` task depends on the result of the `super.test`.

**Reply by ex0ns** [reply]
It makes sense now that you explain it like that, what is pretty confusing is how changing `super.test()` to `super.test()()` could have an impact on the "previous line".

As always, thanks for the explanation
